File: Doubly_linked_list/python/doubly_linked_list.py
class Node:
    def __init__(self, val):
        self.prev = None  # reference to the previous node
        self.elem = val   # value stored in the node
        self.next = None  # reference to the next node


class DoublyLinkedList:
    def __init__(self):
        self.head = None  # head of the list

    # Converts a list to a doubly linked list
    def arr_to_list(self, arr):
        if not arr:
            return
        self.head = Node(arr[0])
        tail = self.head

        for i in range(1, len(arr)):
            new_node = Node(arr[i])
            tail.next = new_node
            new_node.prev = tail
            tail = new_node

    # Returns the node at a specific index
    def node_at(self, idx):
        if idx < 0:
            return None
        temp = self.head
        count = 0
        while temp:
            if count == idx:
                return temp
            count += 1
            temp = temp.next
        return None

    # Inserts a node with value `val` at index `idx`
    def insert(self, idx, val):
        new_node = Node(val)
        if idx == 0:
            new_node.next = self.head
            if self.head:
                self.head.prev = new_node
            self.head = new_node
        else:
            prev_node = self.node_at(idx - 1)
            if prev_node:
                s = prev_node.next
                prev_node.next = new_node
                new_node.prev = prev_node
                new_node.next = s
                if s:
                    s.prev = new_node

    # Removes the node at index `idx`
    def remove(self, idx):
        if idx == 0 and self.head:
            self.head = self.head.next
            if self.head:
                self.head.prev = None
        else:
            prev_node = self.node_at(idx - 1)
            if prev_node and prev_node.next:
                s = prev_node.next.next
                prev_node.next = s
                if s:
                    s.prev = prev_node

    # Prints the list in reverse order
    def reverse_print(self):
        temp = self.head
        if not temp:
            return

        while temp.next:
            temp = temp.next

        while temp:
            print(temp.elem)
            temp = temp.prev


# Testing the doubly linked list
if __name__ == "__main__":
    dll = DoublyLinkedList()

    # Create a doubly linked list from array
    arr = [10, 20, 30, 40, 50]
    dll.arr_to_list(arr)
    print("Original list (reverse print):")
    dll.reverse_print()

    # Insert at the beginning
    dll.insert(0, 5)
    print("\nAfter inserting 5 at index 0:")
    dll.reverse_print()

    # Insert in the middle
    dll.insert(3, 25)
    print("\nAfter inserting 25 at index 3:")
    dll.reverse_print()

    # Insert at the end
    dll.insert(7, 60)
    print("\nAfter inserting 60 at the end:")
    dll.reverse_print()

    # Remove head
    dll.remove(0)
    print("\nAfter removing element at index 0:")
    dll.reverse_print()

    # Remove a middle element
    dll.remove(3)
    print("\nAfter removing element at index 3:")
    dll.reverse_print()

    # Remove last element
    dll.remove(5)
    print("\nAfter removing last element:")
    dll.reverse_print()

File: Doubly_linked_list/java/doubly_linked_list.java
// Node class represents an element in the doubly linked list
class Node {
    Node prev; // reference to the previous node
    int elem; // value stored in the node
    Node next; // reference to the next node

    // Constructor to create a node with a given value
    Node(int val) {
        this.elem = val;
        this.prev = null;
        this.next = null;
    }
}

// DoublyLinkedList class implements the linked list operations
class DoublyLinkedList {
    Node head; // head of the list

    // Converts an array to a doubly linked list
    public void arrToList(int[] arr) {
        this.head = new Node(arr[0]); // initialize head with the first element
        Node tail = this.head; // maintain a tail pointer for appending

        // Loop through remaining elements and add to the list
        for (int i = 1; i < arr.length; i++) {
            Node newNode = new Node(arr[i]); // create a new node
            tail.next = newNode; // link current tail to new node
            newNode.prev = tail; // link new node back to current tail
            tail = newNode; // update tail to new node
        }
    }

    // Returns the node at a specific index
    public Node nodeAt(int idx) {
        if (idx < 0) {
            return null; // invalid index
        }
        Node temp = this.head; // start from head
        int count = 0;

        // Traverse list to find the desired index
        while (temp != null) {
            if (count == idx) {
                return temp; // found node
            }
            count += 1;
            temp = temp.next;
        }

        return null; // index out of bounds
    }

    // Inserts a node with value `val` at index `idx`
    public void insert(int idx, int val) {
        Node newNode = new Node(val); // create a new node

        if (idx == 0) {
            // Inserting at the beginning
            newNode.next = this.head; // new node points to current head
            this.head.prev = newNode; // old head points back to new node
            this.head = newNode; // update head to new node
        } else {
            Node prevNode = nodeAt(idx - 1); // get node before insertion point
            if (prevNode != null) {
                Node s = prevNode.next; // node that will come after the new node
                prevNode.next = newNode; // link previous node to new node
                newNode.prev = prevNode; // link new node back to previous

                newNode.next = s; // link new node to next node
                if (s != null) {
                    s.prev = newNode; // link next node back to new node
                }
            }
        }
    }

    // Removes the node at index `idx`
    public void remove(int idx) {
        if (idx == 0) {
            // Remove the head node
            this.head = this.head.next; // move head to next
            if (this.head != null) {
                this.head.prev = null; // new head has no previous
            }
        } else {
            Node prevNode = nodeAt(idx - 1); // get node before the one to remove
            if (prevNode != null && prevNode.next != null) {
                Node s = prevNode.next.next; // get node after the one being removed
                prevNode.next = s; // link previous node to s
                if (s != null) {
                    s.prev = prevNode; // link s back to previous node
                }
            }
        }
    }

    // Prints the list in reverse order
    public void reversePrint() {
        Node temp = this.head;

        // Go to the last node
        while (temp.next != null) {
            temp = temp.next;
        }

        // Traverse backward and print elements
        while (temp != null) {
            System.out.println(temp.elem);
            temp = temp.prev;
        }
    }
}

// Main class to test the DoublyLinkedList (currently empty)
public class doubly_linked_list {
    public static void main(String[] args) {
        DoublyLinkedList list = new DoublyLinkedList();

        // Create a doubly linked list from array
        int[] arr = { 10, 20, 30, 40, 50 };
        list.arrToList(arr);
        System.out.println("Original list (reverse print):");
        list.reversePrint(); // Should print: 50, 40, 30, 20, 10

        // Insert at the beginning
        list.insert(0, 5);
        System.out.println("\nAfter inserting 5 at index 0:");
        list.reversePrint(); // Should print: 50, 40, 30, 20, 10, 5

        // Insert in the middle
        list.insert(3, 25);
        System.out.println("\nAfter inserting 25 at index 3:");
        list.reversePrint(); // Should print: 50, 40, 30, 25, 20, 10, 5

        // Insert at the end
        list.insert(7, 60);
        System.out.println("\nAfter inserting 60 at the end:");
        list.reversePrint(); // Should print: 60, 50, 40, 30, 25, 20, 10, 5

        // Remove head
        list.remove(0);
        System.out.println("\nAfter removing element at index 0:");
        list.reversePrint(); // Should not contain 5

        // Remove a middle element
        list.remove(3);
        System.out.println("\nAfter removing element at index 3:");
        list.reversePrint(); // Should not contain 30

        // Remove last element
        list.remove(5);
        System.out.println("\nAfter removing last element:");
        list.reversePrint(); // Should not contain 60
    }
}
